### ISSUE TITLE
Custom mods in components

### ISSUE THREAD

**Issue by awinogradov** [bem/bem-components#1687]
Сейчас есть такая проблема, что модификаторы и атрибуты прокидываются только те, что описаны в шаблоне. Например [тут](https://github.com/bem/bem-components/blob/v2/common.blocks/radio-group/radio-group.bemhtml#L20), [тут](https://github.com/bem/bem-components/blob/v2/common.blocks/checkbox/checkbox.bemhtml#L15) или тоже самое с атрибтами [тут](https://github.com/bem/bem-components/blob/v2/common.blocks/input/__control/input__control.bemhtml#L6). Это приводит к огромным проблемам при расширении компонент сверху. Чтобы добавить хотя бы один модификатор или атрибут приходится полностью копировать шаблон компонента, что приводит к необходимости следить за тем, что с этим шаблоном в компонентах, дабы не сломалось чего. Так задумано и ничего не поделать или можно подумать над тем как бы разрешить прокидывание модификаторов?

**Comment by tadatuta** [comment]
Так задумано (см. https://ru.bem.info/libs/principles/#Разделение-на-приватный-и-публичный-api).

Идея в том, что чем больше разрешаешь, тем меньше остается пространства для рефакторинга в рамках минорных версий. Если разрешать все, то любое малейшее изменение кому-нибудь что-нибудь будет ломать.

**Comment by awinogradov** [comment]
Проблема ясна, но получается, что мне надо своей командой мейнтейнить блоки из компонентов или строить свои, чтобы избежать рассинхрона и краха в один прекрасный день. Тут баланс нужен. Может доп моды для элементов блока чтобы экстендить? Что-нибудь?)

**Comment by awinogradov** [comment]
Можно даже с подчеркиванием моды) Как все любят)

**Comment by tadatuta** [comment]
Так а что сейчас мешает экстендить в шаблонах?

Для первого кейса

``` js
block('radio-group').def()(
    return applyNext({ _inGroup: true });
);

block('radio').match(function(){ return this._inGroup; }).def()(function() {
    return this.extend(this.mods, { ololo: 'trololo' });
});
```

Для второго

``` js
block('checkbox').elem('control').def()(function() {
    // ...
});
```

Для третьего

``` js
block('input').elem('control').attrs()(function() {
    return this.extend(applyNext(), { ololo: 'trololo' });
});
```

**Comment by awinogradov** [comment]
Окай, а с [этим](https://github.com/bem/bem-components/blob/v2/common.blocks/checkbox/_type/checkbox_type_button.bemhtml#L8) будет еще знание о модификаторе внешенего блока? Кажется это может к лапше из переменных в дереве привести. Или мне кажется?

**Comment by tadatuta** [comment]
Знание нужно будет пробросить как в первом кейсе.

Очевидно, что отсутствие готового API потребует больше кода для переопределений, но предполагается, что такие дополнения все-таки требуются далеко не всем и далеко не всегда.

Если же мы упустили какой-то конкретный очень частый кейс, то создай для него отдельный issue, обсудим.

**Comment by awinogradov** [comment]
Доопределения вообще никому не требуются, если предполагается, что компоненты это просто пример библиотеки и мол если хотите кастом, то пилите свои. Само собой у всех своя специфика и вот тут свои проблемы лью и думаю: вдруг выгорит и поудобнее станет)

**Comment by tadatuta** [comment]
Нет, это не пример, а скорее база. Т.е. ожидается, что для абсолютного большинства кейсов текущего API должно быть достаточно, чтобы мочь реализовать свои наследованием или композицией с учетом того, что:
1. Дизайн можно отстрелить и реализовать свой
2. Любой абстрактный блок из `bem-components` можно обернуть в свой конкретный (либо отнаследоваться), где и реализовать кастомное API
3. Все можно заинжектится с помощью шаблонов, но да, это будет не так удобно, как если бы API было полностью открытым

**Comment by awinogradov** [comment]
И шоже нам делать?)

**Comment by tadatuta** [comment]
Таки писать про конкретные частые кейсы, которые мы не учли и про кейсы, которые нельзя решить композицией/наследованием/переопределением, если они редкие (ну, т.е. делать частые штуки должно быть просто, а редкие — возможно).
